src/sources/util/websocket/source.rs

```diff
@@ -0,0 +1,238 @@
+                                        Some(WebSocketEvent{
+                                            name: DEFAULT_SOURCE_NAME.to_owned(),
+                                            payload: log_evt.value().to_string(),
+                                            log_namespace: &params.log_namespace,
+                                        })
+                                    } else {
+                                        warn!("Decoded unsupported event: {:?}", e);
+                                        None
+                                    }
+                                }).map(|evt| async {
+                                    handle_text_message(&mut out.clone(), evt, config.uri.clone()).await
+                                }).ok_or(())?.await;
+                                Ok::<(), ()>(())
+                            }).map_err(|err| {error!("Failed to process binary message: {}", err);}) {
+                                Ok(_) => Ok(()),
+                                Err(e) => {
+                                    error!("Failed to send binary message: {:?}", e);
+                                    Ok(())
+                                }
+                            }
+                    },
+
+                    Ok(Message::Close(_)) => {
+                        info!("Received message: connection closed from server");
+                        Err(WsError::ConnectionClosed)
+                    },
+
+                    Ok(Message::Frame(_)) => {
+                        warn!("Unsupported message type received: frame");
+                        Ok(())
+                    },
+
+                    Err(e) => Err(e),
+                }
+            }
+        };
+
+        if let Err(error) = result {
+            if is_closed(&error) {
+                emit!(WsConnectionShutdown);
+                return Err(());
+            } else {
+                emit!(WsConnectionError { error });
+                (*ws_sink, *ws_source) = create_sink_and_stream(params.connector.clone()).await;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+async fn create_sink_and_stream(
+    connector: WebSocketConnector,
+) -> (
+    impl Sink<Message, Error = WsError>,
+    impl Stream<Item = Result<Message, WsError>>,
+) {
+    let ws_stream = connector.connect_backoff().await;
+    ws_stream.split()
+}
+
+fn check_received_pong_time(
+    ping_timeout: Option<NonZeroU64>,
+    last_pong: Instant,
+) -> Result<(), WsError> {
+    if let Some(ping_timeout) = ping_timeout {
+        if last_pong.elapsed() > Duration::from_secs(ping_timeout.into()) {
+            return Err(WsError::Io(io::Error::new(
+                io::ErrorKind::TimedOut,
+                "Pong not received in time",
+            )));
+        }
+    }
+
+    Ok(())
+}
+
+async fn handle_text_message<'a>(
+    out: &mut SourceSender,
+    msg: WebSocketEvent<'a>,
+    endpoint: String,
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏I think this can be passed in as a reference to avoid the clone, and the `WsMessageReceived` can specify the lifetime like 

```rust
pub uri: &'a str,
```